### ISSUE TITLE
computed_field with MissingType in return type: PydanticSerializationError: Unable to serialize unknown type: <MissingType> 

### ISSUE THREAD

**Issue by thibaut-lo** [joshorr/pydantic-partials#28]
Hi - this is a very nice lib! Just reporting something I saw

```python
from pydantic_partials import PartialModel, MissingType, Missing

from pydantic import computed_field

class Test(PartialModel):
    a: int

    @computed_field
    @property
    def b(self) -> int | MissingType:
          if a < 2:
              return Missing
          return a

t = Test(a=1)
t.model_dump_json()
```

raises
```txt
.local/lib/python3.11/site-packages/pydantic/main.py", line 415, in model_dump_json
    return self.__pydantic_serializer__.to_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.PydanticSerializationError: Unable to serialize unknown type: <class 'pydantic_partials.sentinels.MissingType'>
```

Instead, I would expect it to drop the computed_fields.

This is nit.

## Versions

pydantic-partials: 1.1.0
pydantic: 2.8.2

**Comment by joshorr** [comment]
@thibaut-lo Thank you for the report, very much appreciated!

I'll try and take a look at this be end of the week, see if I can get it fixed.  I don't think I've tested the library with computed fields, so it's probably something I am not taking into account that I should have.  I'll be sure to add a unit-test when I fix it.

Thank you!

**Comment by joshorr** [comment]
Thinking about it for a few minutes, it's probably related to the biggest limitation I have right now with Pedantic. I can't raise a `PydanticOmit` in a serializer without Pedantic producing an error (See https://github.com/pydantic/pydantic/discussions/5461#discussioncomment-7411977).

I actually delete/prevent `Missing` values from being assigned to attributes right now as a work-around.  A computed field however can return a value and I can't prevent it by deleting the attribute (since it's computed field property function).

When I get some time later this week to look into this, I'll have to see if I can figure out a workaround for this use-case, until there is some way to tell Pedantic not to serialize something from the serializer method.

**Comment by joshorr** [comment]
@thibaut-lo Sorry, I have not had a chance to see if there is a workaround (had a family emergency over last weekend).

There might not be any good workaround, until `PydanticOmit` can be used in a serializer (see above).  The Pydantic people seem interested in eventually doing that, but it's not a high priority for them at the moment.

Perhaps I can override the `model_dump` method and try to deal with it there.  It won't fix `model_dump_json`, and I would have to recursively look the dump to find the `Missing` values and remove them.

So it's not ideal and a bit messy, but it may support your use-case well enough for now to be worth doing.  Like I said, I'll see if I can do this later this week.  We will see.

**Comment by joshorr** [comment]
@thibaut-lo The good news is it sounds like there is a new feature I the next release of Pydantic (v2.11) that will allow one to conditionally exclude fields via a user-provided function.  I could probably use this to do what is needed.  As soon as the `v2.11` is out, I'll attempt to utilize that new feature and get your use-case working.

Today, I was actually going to try modifying the base-model and override `model_dump` to at least get that specific function working with your use case.  But in light of this soon to be released feature in v2.11 of Pydantic, I'm going to wait for that and see if that new feature is good enough for your use case here.

## Comments About Code

If it turns out that `v2.11` can't be used for your use case, I'll serasouly consider adding your code (with some tweaks) to the PartialModel.

### Just a few side comments about the code you provided (in case it's helpful):

The hard thing about `model_serializer` is that the super-class's version won't be called if a subclass does its own `model_serializer`.  This is one of the reasons I decided to not go this route in the first place.  Also, there could be private fields (so the `self.__dict__.keys()` might include other attributes that don't matter).  You would be better off using the `ModelCls.model_fields` property to get the Pydantic-aware fields and use that key-list instead.

**Comment by thibaut-lo** [comment]
Forgot to say thanks for both feedbacks! The new `exclude-if` pydantic feature sounds promising! Thanks again.

**Comment by joshorr** [comment]
Update: Still waiting on `v2.11` of Pydantic.  I am keeping an eye on it.

**Comment by joshorr** [comment]
@thibaut-lo Update: Pydantic v2.11 came out a couple of days ago.  Unfortunately, it does not seem to have the `exclude_if` feature.  It did seem like they were going to do it, as the feature is still on the Pydantic v2.11, and I see a comment on both the core and main Pydantic library (https://github.com/pydantic/pydantic/pull/10961#issuecomment-2654622257).

I left a comment on both PR's asking when it might come out.  Sorry about the delay in finishing this feature 🙁

**Comment by joshorr** [comment]
@thibaut-lo I decided to reopen this because even though version 3 supports the computed fields + Missing, version 4 does not. So for now, you can stick to version 3 if you want the automatic partials functionality this library has.  Pydantic will be changed soon to allow me to re-support this feature, and then I'll close this issue once that happens.

## Details

This is due to it not being wise to use the `MISSING` experimental feature just yet as it broke mypy (https://github.com/joshorr/pydantic-partials/issues/42) plus had IDE/Editor issues.

In v4 I use the new `exclude_if` feature from Pydantic.  That feature does not currently support computed fields.  I opened an issue here (https://github.com/pydantic/pydantic/issues/12690)  and it looks like they are going to support this for computed fields in the near future (https://github.com/pydantic/pydantic/pull/12677#discussion_r2676206759).  

As soon as they do, it should just start working again.  As soon as Pydantic releases this change, I'll confirm it works and re-enable my unit test that tests for computed-fields + missing.

**Comment by joshorr** [comment]
@thibaut-lo It looks like the release a new version of Pydantic a week or so back (v2.13).

I'll take a look this week and see if the exclude-if feature works with the computed-fields now and if so get it published.  The work for using this feature with computed fields is almost finished already; I just need to enable it and verify with some unit tests.

**Comment by joshorr** [comment]
@thibaut-lo Looks like they closed my issue (https://github.com/pydantic/pydantic/issues/12690) as "completed" 😃!  So it should work now after I finish my side of the work. I'll get this done this week.

**Comment by joshorr** [comment]
@thibaut-lo I have just released a new pydantic-partials version (v4.1.0) that has full support for returning `Missing` in computed fields.
